From: Chris Renshaw <osm0sis@outlook.com>
Date: Fri, 2 Feb 2018 16:37:54 -0400
Subject: [PATCH] fix time+hush: add missing definitions on Android API 21+

wait3() wrapper from 'sys-wait.h.patch' by Fredrik Fornwall:
https://github.com/termux/termux-packages/tree/master/ndk_patches/

issetugid is required for hush via glob.c; add the hacky pre-API 21 definition back

---
 include/platform.h | 17 +++++++++++++++++
 libbb/platform.c   |  8 ++++++++
 2 files changed, 25 insertions(+)

diff --git a/include/platform.h b/include/platform.h
index cfedef0fc..5750f9575 100644
--- a/include/platform.h
+++ b/include/platform.h
@@ -388,6 +388,8 @@ typedef unsigned smalluint;
 #define HAVE_NET_ETHERNET_H 1
 #define HAVE_SYS_STATFS_H 1
 #define HAVE_PRINTF_PERCENTM 1
+#define HAVE_WAIT3 1
+#define HAVE_ISSETUGID 1
 
 #if defined(__UCLIBC__)
 # if UCLIBC_VERSION < KERNEL_VERSION(0, 9, 32)
@@ -489,10 +491,15 @@ typedef unsigned smalluint;
 # else
    /* ANDROID >= 21 has standard dprintf */
 # endif
+# if __ANDROID_API__ > 18
+#  undef HAVE_ISSETUGID
+# endif
 # if __ANDROID_API__ < 21
 #  undef HAVE_TTYNAME_R
 #  undef HAVE_GETLINE
 #  undef HAVE_STPCPY
+# else
+#  undef HAVE_WAIT3
 # endif
 # undef HAVE_MEMPCPY
 # undef HAVE_STRCHRNUL
@@ -580,4 +587,14 @@ extern int vasprintf(char **string_ptr, const char *format, va_list p) FAST_FUNC
 extern ssize_t getline(char **lineptr, size_t *n, FILE *stream) FAST_FUNC;
 #endif
 
+#ifndef HAVE_WAIT3
+/* Wrap wait3() to wait4() for libc implementations without (e.g. Bionic on ANDROID >= 21) */
+# include <sys/wait.h> /* for rusage */
+static pid_t wait3(int* status, int options, struct rusage* rusage) { return wait4(-1, status, options, rusage); }
+#endif
+
+#ifndef HAVE_ISSETUGID
+extern int issetugid(void) FAST_FUNC;
+#endif
+
 #endif
diff --git a/libbb/platform.c b/libbb/platform.c
index 03bbb798b..a5b60074a 100644
--- a/libbb/platform.c
+++ b/libbb/platform.c
@@ -213,3 +213,11 @@ int ttyname_r(int fd, char *buf, size_t buflen)
 	return 0;
 }
 #endif
+
+#ifndef HAVE_ISSETUGID
+int issetugid(void)
+{
+	/* for Bionic, this is sufficient */
+	return 0;
+}
+#endif
-- 
2.12.3

